### ISSUE TITLE
Lamb's Gap Tower processing error

### ISSUE THREAD

**Issue by zcrysler** [MotusDev/Motus-TO-DO#370]
Project 134, can this be re-run? The error says:

Error: Error in rsqlite_send_query(conn@ptr, statement): Expecting a single string value: [type=character; extent=0].

**Comment by Hussell** [comment]
Rerunning this just gives the same error. The log reports "Inf" (infinity) for some times, where it should be seconds. The file has beeper records, which might not be parsing correctly?

**Comment by zcrysler** [comment]
So should the user do something or is this a bug to fix on our end?

**Comment by leberrigan** [comment]
User recently uploaded a new file from the same receiver and got the same error (Job ID: 228117). There have been no successful uploads for this receiver. The DTA file for the latest upload is here: [Lamb Gap.zip](https://github.com/MotusDev/Motus-TO-DO/files/2691037/Lamb.Gap.zip)

Has any progress been made on this issue since Oct?

**Comment by Hussell** [comment]
Lucas: no, this got buried in the list of things to do, so I hadn't looked at it again. Sorry.

John:
https://sgdata.motus.org/status?jobID=228117
https://sgdata.motus.org/status?jobID=218196

I don't understand the ultimate cause here, but from examining the stack dump I see that in ltFindTags.R btrec$mints is a numeric(0), so the sprintf to insert that into an SQL command is returning character(0), which is the proximate cause of the error.

**Comment by jbrzusto** [comment]
### TLDR ###
Parsing of the .DTA file is failing on the 2nd `Antenna Gain` block.  This is due to an encoding mis-match between regex matching and substringing stages resulting from a non-ascii Latin-1 character (`±`) in the .DTA file.  I'm posting the code issue as jbrzusto/motusServer#429.  When the fix is made and the servers restarted, this job can be re-run, and the following checked off:
- [ ] delete records of the offending file from `/sgm/recv/Lotek-D000465.motus`:
```sql
```
- [ ] rerun upload job

### Details ###
```R
## load the stack dump as indicated in the error dialog you get from the status page when.
## clicking on the subjob with the error ("ltFindTags")
> x = readRDS("/sgm/www//errors/00218205.rds")
> names(x).
[1] "rsqlite_send_query(conn@ptr, statement)"                                          
[2] "initialize(value, ...)"                                                           
[3] "initialize(value, ...)"                                                           
[4] "new(\"SQLiteResult\", sql = statement, ptr = rsqlite_send_query(conn@ptr, sta"    
[5] ".local(conn, statement, ...)"                                                     
[6] "is(object, Cl)"                                                                   
[7] "is(object, Cl)"                                                                   
[8] ".valueClassTest(standardGeneric(\"dbSendQuery\"), \"DBIResult\", \"dbSendQuery\")"
[9] "dbSendQuery(conn, statement, ...)"                                                
[10] "dbGetQuery(src$con, sprintf(\"select not exists (select * from DTAtags as t1"     
[11] "dbGetQuery(src$con, sprintf(\"select not exists (select * from DTAtags as t1"     
[12] "ltFindTags(src, MOTUS_METADB_CACHE, par = paste(ltDefaultFindTagsParams, po"      
[13] "h(j)"    .
## as @Hussell pointed out, the SQL query is faulty, due to `btrec` being empty:
> x[[12]]$btrec
[1] relboot mints
<0 rows> (or 0-length row.names)
## This is supposed to be a record of the boot time of the receiver (according to
## its own clock, as recorded by an "Environment History" group in the .DTA file
## So try parse the .DTA file and see what breaks:
> options(error=recover)
> library(motusServer).
## skipping diagnostics
> x = readDTA("/sgm/uploads/2398/2398_2018-10-16T18-15-23.629_Lamb_Gap.DTA")
Error in `data.frame`(tab, , 2) : undefined columns selected
Enter a frame number, or 0 to exit
1: readDTA("2398_2018-10-16T18-15-23.629_Lamb_Gap.DTA")
2: tab[, 2]
3: `data.frame`(tab, , 2)

Selection:1
Called from: readDTA("2398_2018-10-16T18-15-23.629_Lamb_Gap.DTA")
Browse[1]> ls()
 [1] "boottimes"          "clen"               "codeset"           
 [4] "con"                "cstart"             "date.format"       
 [7] "filename"           "freq.tab"           "gain.tab"          
[10] "ip"                 "lines"              "model"             
[13] "newlines"           "numLines"           "parts"             
[16] "piece.lines.before" "piece.name"         "pieces"            
[19] "res"                "serno"              "site_code"         
[22] "tab"                "tags"
Browse[1]> tab
      X75
2      75
3      75
4      75
       NA
Maste  NA
Browse[1]> piece.name
[1] "antenna_gain"
## so it's weird; the above table (`tab`) is supposed to be the table of antenna gain settings.
## Looking at the names of all pieces:
Browse[1]> names(pieces)
 [1] "model"        "boottime"     "serial_no"    "code_set"     "antenna_gain"
 [6] "model"        "boottime"     "serial_no"    "code_set"     "antenna_gain"
[11] "active_scan"  "model"        "boottime"     "serial_no"    "code_set"    
[16] "antenna_gain" "active_scan"  "model"        "boottime"     "serial_no"   
[21] "code_set"     "antenna_gain" "active_scan"  "model"        "boottime"    
[26] "serial_no"    "code_set"     "antenna_gain" "active_scan"  "model"       
[31] "boottime"     "serial_no"    "code_set"     "antenna_gain" "active_scan" 
[36] "model"        "boottime"     "serial_no"    "code_set"     "antenna_gain"
[41] "active_scan"  "model"        "boottime"     "serial_no"    "code_set"    
[46] "antenna_gain" "active_scan"  "model"        "boottime"     "serial_no"   
[51] "code_set"     "antenna_gain" "active_scan"  "model"        "boottime"    
[56] "serial_no"    "code_set"     "antenna_gain" "active_scan"  "model"       
[61] "boottime"     "serial_no"    "code_set"     "antenna_gain" "active_scan" 
[66] "model"        "boottime"     "serial_no"    "code_set"     "antenna_gain"
[71] "active_scan"  "id_only"
## we see there are two "antenna_gain" pieces; which one caused the problem?
Browse[1]> ip
[1] 10
Browse[1]> pieces[5]
                                                              antenna_gain 
"   1          75\n   2          75\n   3          75\n   4          75\n" 
Browse[1]> pieces[10]
                                                               antenna_gain 
"        75\n   2          75\n   3          75\n   4          75\n\nMaste" 
```
So why did the two antenna gain blocks parse differently?  They are identical in the .DTA file:
```
offset 000000f2:                                          offset 00000550:
524f 5550 3a0d 0a41 6e74 656e 6e61 2020  ROUP:..Antenna   524f 5550 3a0d 0a41 6e74 656e 6e61 2020  ROUP:..Antenna
2020 2020 4761 696e 0d0a 2020 2031 2020      Gain..   1   2020 2020 4761 696e 0d0a 2020 2031 2020      Gain..   1
2020 2020 2020 2020 3735 0d0a 2020 2032          75..   2 2020 2020 2020 2020 3735 0d0a 2020 2032          75..   2
2020 2020 2020 2020 2020 3735 0d0a 2020            75..   2020 2020 2020 2020 2020 3735 0d0a 2020            75..
2033 2020 2020 2020 2020 2020 3735 0d0a   3          75.. 2033 2020 2020 2020 2020 2020 3735 0d0a   3          75..
2020 2034 2020 2020 2020 2020 2020 3735     4          75 2020 2034 2020 2020 2020 2020 2020 3735     4          75
0d0a 0d0a 4d61 7374 6572 2041 6e74       ....Master Ant   0d0a 0d0a 4d61 7374 6572 2041 6e74       ....Master Ant
```
It appears the starting offset used to substring the parsed `antenna_gain` piece is wrong, and this is confirmed by looking at the first two `model` pieces:
```R
Browse[1]> pieces[c(1,6)]
model    model
"SRX800" " / 800" 
```
These come from identical lines:
```
SRX800 / 800D Information:
SRX800 / 800D Information:
```
So somewhere between the first `antenna_gain` and the second `model` pieces, something is throwing off substring offsets. The culprit appears to be the two `±` characters present as single `0xB1` bytes (their Latin-1 encoding) in the original file:
```
Active scan_table:
CHANNEL  FREQUENCY   STATUS     TYPE         BPM WINDOWS
   2      166.849    Skipped    Beeper Tag   30±10
   1      166.380    Enabled    Beeper Tag   30±10
```
or in hex:
```
000002c0: 6564 2020 2020 4265 6570 6572 2054 6167  ed    Beeper Tag
000002d0: 2020 2033 30b1 3130 0d0a 2020 2031 2020     30.10..   1  
```
Indeed, when I resave a copy of the file replacing `±` with the ascii triplet `+/-`, the DTA file
is read correctly.

**Comment by Hussell** [comment]
So the ultimate cause is... the file has beeper records which aren't being parsed correctly? [Ha!](https://github.com/MotusDev/Motus-TO-DO/issues/370#issuecomment-425440872)

Thanks, John, I really appreciate your thorough write-ups of your debugging process, both on this and other issues.
